File: 5_file_code/node.js/chat_client.js
const net = require('net');
const readline = require('readline');

const PORT = 9999;
const HOST = '127.0.0.1';

const client = new net.Socket();

// Thiết lập giao diện nhập liệu từ bàn phím
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

client.connect(PORT, HOST, () => {
    console.log('------------------------------------------------');
    console.log(`Đã kết nối tới Chat Server tại ${HOST}:${PORT}`);
    console.log("Gõ tin nhắn và nhấn Enter để gửi. Gõ 'exit' để thoát.");
    console.log('------------------------------------------------');
    
    // Bắt đầu cho phép nhập
    promptUser();
});

// Sự kiện nhận dữ liệu từ Server
client.on('data', (data) => {
    // Xóa dòng hiện tại (để tin nhắn đến không bị chèn vào dòng đang gõ)
    process.stdout.clearLine();
    process.stdout.cursorTo(0);
    
    console.log(data.toString().trim());
    
    // Hiện lại dấu nhắc để nhập tiếp
    promptUser();
});

client.on('close', () => {
    console.log('\n[Mất kết nối với Server]');
    process.exit(0);
});

client.on('error', (err) => {
    console.log(`\n[LỖI KẾT NỐI]: ${err.message}`);
});

// Hàm xử lý nhập liệu
function promptUser() {
    rl.question('[Bạn]: ', (message) => {
        if (message.toLowerCase() === 'exit') {
            client.end();
            rl.close();
            process.exit(0);
        } else {
            // Gửi tin nhắn lên server
            client.write(message);
            // Tiếp tục vòng lặp nhập liệu (Nhưng chờ readline xử lý xong)
            // Lưu ý: rl.question sẽ tự lock luồng nhập cho đến khi Enter, 
            // nên ta gọi lại promptUser bên trong callback này là đúng logic.
        }
    });
}

File: 5_file_code/golang/chat_client.go
package main

import (
	"bufio"
	"fmt"
	"net"
	"os"
	"strings"
)

const (
	HOST = "localhost"
	PORT = "9999"
	TYPE = "tcp"
)

func main() {
	conn, err := net.Dial(TYPE, HOST+":"+PORT)
	if err != nil {
		fmt.Println("Không thể kết nối tới Server:", err)
		return
	}
	defer conn.Close()

	// --- TỐI ƯU HÓA TCP ---
	if tcpConn, ok := conn.(*net.TCPConn); ok {
		tcpConn.SetNoDelay(true) // Gửi tin tức thì
	}

	fmt.Println("------------------------------------------------")
	fmt.Println("Đã kết nối tới Go Chat Server!")
	fmt.Println("Gõ tin nhắn và nhấn Enter. Gõ 'exit' để thoát.")
	fmt.Println("------------------------------------------------")

	// --- LUỒNG 1: NHẬN TIN (Chạy ngầm) ---
	go func() {
		scanner := bufio.NewScanner(conn)
		for scanner.Scan() {
			fmt.Println(scanner.Text())
		}
		// Nếu vòng lặp kết thúc nghĩa là mất kết nối
		fmt.Println("\n[Mất kết nối với Server]")
		os.Exit(0)
	}()

	// --- LUỒNG 2 (CHÍNH): GỬI TIN ---
	consoleScanner := bufio.NewScanner(os.Stdin)
	fmt.Print("[Bạn]: ") // Dấu nhắc ban đầu

	for consoleScanner.Scan() {
		text := consoleScanner.Text()

		if strings.ToLower(text) == "exit" {
			fmt.Println("Đang thoát...")
			break
		}

		// Gửi lên server (thêm \n để server biết hết câu)
		fmt.Fprintf(conn, "%s\n", text)

		// In lại dấu nhắc
		// Lưu ý: Do bất đồng bộ, dấu nhắc này có thể bị trôi khi có tin nhắn đến
		// nhưng đây là cách đơn giản nhất cho Demo.
	}
}
